### ISSUE TITLE
[Feature Request] List only stopped/started services

### ISSUE THREAD

**Issue by xgqt** [OpenRC/openrc#612]
Would it be possible to add --stopped and --started options to rc-status?
We already have --crashed.

**Comment by Higgs1** [comment]
+1 I too would like this.

In the code, there's a nifty functioned named [`rc_services_in_state`](https://github.com/OpenRC/openrc/blob/-/src/librc/librc.c#L1005-L1036) which is already utilized for `--crashed`. The possible service states are ([source](https://github.com/OpenRC/openrc/blob/-/src/librc/librc.c#L1005-L1036)):

* started
* stopped
* starting
* stopping
* inactive
* wasinactive
* hotplugged
* failed
* scheduled
* crashed

I don't mean to offend by hijacking your FR, but all of those states could hypothetically have their own filter argument, especially because they already seem to have English-people-friendly names. (e.g. `--started`, `--stopped`, `--starting`, etc.

If multiple filters are specified... well, it wouldn't make any sense to AND them together, because a service can only be in a single state. They could be OR'd together instead.

Then again I might just be talking nonsense. This is just my opinion.

**Comment by xgqt** [comment]
> I don't mean to offend by hijacking your FR, but all of those states could hypothetically have their own filter argument, especially because they already seem to have English-people-friendly names. (e.g. --started, --stopped, --starting, etc.

Haha, I do not mind. I want to have in any form; this, that or even different.

**Comment by N-R-K** [comment]
> because a service can only be in a single state.

Not really true. Not all the states are mutually exclusive:

https://github.com/OpenRC/openrc/blob/72b1e8cd0c08dee3efb0954298453ee7c56f2037/src/librc/rc.h.in#L212-L216

This is why `rc-status -c` gets all the STARTED services and then only prints the crashed ones:

https://github.com/OpenRC/openrc/blob/72b1e8cd0c08dee3efb0954298453ee7c56f2037/src/rc-status/rc-status.c#L281-L287

> Would it be possible to add --stopped and --started options to rc-status?

@xgqt Should it also include services that may have crashed?

**Comment by xgqt** [comment]
@N-R-K 

> Should it also include services that may have crashed?

No, I dont think --started and --stopped  should include the crashed ones.

**Comment by N-R-K** [comment]
> No, I dont think --started and --stopped should include the crashed ones.

Okay, I'm assuming you don't want services in failed state either for `--started` output.

And on `--stopped` are you expecting services which were started and then stopped? Or all existing services which are preset but not running?

**Comment by xgqt** [comment]
@N-R-K 

> And on `--stopped` are you expecting services which were started and then stopped? 

yes

>  Or all existing services which are not running?

I would expect services that are enabled but stopped.

So for example if I just installed nginx and ran `rc-service add nginx default`, then `rc-status --stopped` would show only nginx.

**Comment by navi-desu** [comment]
we can't really track "started-then-stopped", the best would be "in the current runlevel but stopped" -- which is basically filtering what rc-status already does

**Comment by N-R-K** [comment]
Since the OP mentioned `--crashed` which ignores runlevels and prints *all crashed* services, I started with the assumption that that's what OP wanted with `--started` and `--stopped` as well. But looks like that's not the case.

Okay, so let's start over, here are some questions:

1. Should `--started` include crashed and failed services? (Ans: NO)
2. Should `--started` (and `--stopped`) print all started/stopped services? Only in the current runlevel? In the runlevels specified on the command line as arguments?
3. Should it output only the service names on each line (similar to `--crashed`)? Or should it do fancy output (similar to running `rc-status` without any argument)?

**Comment by navi-desu** [comment]
what would make sense to me, is the --$state flags to basically just, filter whatever output, so, for example:

```
 $ rc-status --user                                                                                                                                               
Runlevel: gui
 dunst                                                 [  started 15:04:58 (0) ]
 xdg-desktop-portal-wlr                                            [  crashed  ]
Stacked Runlevel: default
 pipewire-pulse                                        [  started 15:05:01 (0) ]
 wireplumber                                           [  started 15:05:01 (0) ]
 mpd                                                               [  stopped  ]
Dynamic Runlevel: hotplugged
Dynamic Runlevel: needed/wanted
 dbus                                                  [  started 15:05:01 (0) ]
 pipewire                                              [  started 15:05:01 (0) ]
Dynamic Runlevel: manual
```

with --started, it would print
```
Runlevel: gui
 dunst                                                 [  started 15:04:58 (0) ]
Stacked Runlevel: default
 pipewire-pulse                                        [  started 15:05:01 (0) ]
 wireplumber                                           [  started 15:05:01 (0) ]
Dynamic Runlevel: hotplugged
Dynamic Runlevel: needed/wanted
 dbus                                                  [  started 15:05:01 (0) ]
 pipewire                                              [  started 15:05:01 (0) ]
Dynamic Runlevel: manual
```

and --crashed
```
Runlevel: gui
 xdg-desktop-portal-wlr                                            [  crashed  ]
Stacked Runlevel: default
Dynamic Runlevel: hotplugged
Dynamic Runlevel: needed/wanted
Dynamic Runlevel: manual
```

same with stopped and etc, then a "plain" output that just prints lines could be had by either detecting if stdout is not a tty, or, a --format flag (similarly to how we have -f ini). this could also allow for multiple flags to happen at once, say --started --crashed would filter everything that isn't those two flags

note that this would be a breaking change wrt to the default --crashed, but it makes more sense, imo

wdyt?

**Comment by N-R-K** [comment]
> instead of adding 5 different flags, maybe we'd have a single flag like `--state` that takes an argument?

Yes, I've been thinking the same. It also avoids having to break compatibility with `--crash`. `--crash` can keep working the way it currently does and `--state "crash"` will use the newer behavior.

**Comment by xgqt** [comment]
> what would make sense to me, is the --$state flags to basically just, filter whatever output, so, for example:

yup

> also alternatively, instead of adding 5 different flags, maybe we'd have a single flag like --state that takes an argument?

sounds good to me

> note that this would be a breaking change wrt to the default --crashed, but it makes more sense, imo

Wouldn't it be possible to have both --state and --crashed as-was?

> can keep working the way it currently does and --state "crash" will use the newer behavior.

that looks ok

**Comment by navi-desu** [comment]
> Wouldn't it be possible to have both --state and --crashed as-was?

yeah (my post is a bit out of order, i meant it like "if we don't add state, it would break...")

since states are just bitflags, i suggest `--in-state` or short-hand `-i`, then something like (`--state` and `-s` would've been nice, but it's already taken)

`rc-status -i crashed -i failed -i stopped` works as intended and is easy to write, then we just bitwise-and them, and check before printing (would need to also work with `--servicelist` for all services too)

**Comment by N-R-K** [comment]
I have a [WIP branch](https://github.com/N-R-K/openrc/tree/rc-status-state-print) which implements `--in-state` flag. Please try it out and report if it behaves as expected or not. If it's behaving properly, I can start adding documentation and raise a PR.
